File: problems/arrays/line_through_points/js/line_through_points.js
export const lineThroughPoints = (() => {
  const gcd = (a, b) => {
    while (b !== 0) [a, b] = [b, a % b];

    return a;
  };

  const normalize = (dx, dy) => {
    const gcdVal = gcd(dx, dy);
    dx /= gcdVal;
    dy /= gcdVal;

    if (dx < 0) {
      dx = -dx;
      dy = -dy;
    }
    return [dx, dy];
  };

  const getMaxPoints = (index, points, p1) => {
    const slopes = new Map();
    let duplicate = 1;
    const [x1, y1] = p1;

    for (let j = index + 1; j < points.length; j++) {
      const [x2, y2] = points[j];

      if (x1 === x2 && y1 === y2) {
        duplicate++;
        continue;
      }
      let [dx, dy] = normalize(x2 - x1, y2 - y1);
      const slope = `${dy}/${dx}`;
      slopes.set(slope, (slopes.get(slope) || 0) + 1);
    }
    return (slopes.size ? Math.max(...slopes.values()) : 0) + duplicate;
  };

  return (points) => {
    if (points.length < 3) return points.length;

    let maxPoints = 0;

    for (let i = 0; i < points.length; i++) {
      maxPoints = Math.max(maxPoints, getMaxPoints(i, points, points[i]));
    }
    return maxPoints;
  };
})();

File: problems/arrays/line_through_points/python/line_through_points.py
from collections import defaultdict
from math import gcd


def normalize(dx: int, dy: int) -> tuple[int, int]:
    def compute_gcd(a: int, b: int) -> int:
        return gcd(a, b) if b != 0 else a

    gcd_val = compute_gcd(dx, dy)
    dx //= gcd_val
    dy //= gcd_val
    if dx < 0:
        dx, dy = -dx, -dy

    return dx, dy


def get_max_points(index: int, points: list[list[int]], p1: tuple[int, int]) -> int:
    slopes = defaultdict(int)
    duplicate = 1
    x1, y1 = p1

    for j in range(index + 1, len(points)):
        x2, y2 = points[j]
        if x1 == x2 and y1 == y2:
            duplicate += 1
            continue

        dx, dy = normalize(x2 - x1, y2 - y1)
        slopes[(dx, dy)] += 1

    return max(slopes.values(), default=0) + duplicate


def lineThroughPoints(points: list[list[int]]) -> int:
    if len(points) < 3:
        return len(points)

    max_points = 0
    for i in range(len(points)):
        max_points = max(
            max_points, get_max_points(i, points, [points[i][0], points[i][1]])
        )

    return max_points
